Murloc11/bjs2 84050 users who liked it

bookinisto opened the pull request.

bookinisto committed:
BJS2-84050
сделал :
- два метода для получения списка пользователей кто поставил лайк посту/коментарию
- тесты на эти методы

MyrzakhmetSmagul commented on src/main/java/faang/school/postservice/controller/like/LikeController.java:
```diff
@@ -0,0 +1,79 @@
+        return ResponseEntity.noContent().build();
+    }
+
+    @GetMapping("/posts/{postId}/likes")
```
тоже замечание что и снизу
как вариант так же можешь использовать методы и просто оставить по типу 
get `/likes/posts/{postId}`
post `/likes/posts/{postId}`

MyrzakhmetSmagul commented on src/main/java/faang/school/postservice/service/like/LikeServiceImpl.java:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    @Override
+    public List<UserDto> getListUsersWhoLikesThisPost(long postId) {
```
названия методом можно упростить
getPostLikers(long postId)
findPostLikers(long postId)
getUsersWhoLikePost(long postId)
findUsersWhoLikePost(long postId)
над названиями можно поработать вместе с гпт

bookinisto replied:
ок

MyrzakhmetSmagul commented on src/main/java/faang/school/postservice/service/like/LikeServiceImpl.java:
```diff
@@ -0,0 +1,151 @@
+
+    @Override
+    public List<UserDto> getListUsersWhoLikesThisPost(long postId) {
+        Post post = postRepository.findById(postId)
```
у тебя есть `getRequiredById` который ты создал, он делает тоже самое

bookinisto replied:
ок

MyrzakhmetSmagul commented on src/main/java/faang/school/postservice/service/like/LikeServiceImpl.java:
```diff
@@ -0,0 +1,151 @@
+
+        return likes.stream()
+                .map(like -> userServiceClient.getUser(like.getUserId()))
+                .collect(Collectors.toList());
```
можно заменить на `.toList()`

bookinisto replied:
ok

MyrzakhmetSmagul commented on src/main/java/faang/school/postservice/service/like/LikeServiceImpl.java:
```diff
@@ -0,0 +1,151 @@
+
+    @Override
+    public List<UserDto> getListUsersWhoLikesThisComment(long commentId) {
+        Comment comment = commentRepository.findById(commentId)
```
`getRequiredById`

MyrzakhmetSmagul commented on src/main/java/faang/school/postservice/service/like/LikeServiceImpl.java:
```diff
@@ -0,0 +1,151 @@
+
+        return likes.stream()
+                .map(like -> userServiceClient.getUser(like.getUserId()))
+                .collect(Collectors.toList());
```
`.toList`

RinatBeybutov commented:
Комментарии Мурзахмета не исправлены, сначала нужно поправить их.

Еще нужно поправить конфликт с main веткой

bookinisto commented on src/main/java/faang/school/postservice/service/like/LikeServiceImpl.java:
```diff
@@ -0,0 +1,151 @@
+            throw new IllegalStateException("You already liked this post");
+        }
+
+        if (likeRepository.existsByUserIdAndPostIdOnComments(userId, postId)) {
```
такие условия были поставленны в задаче, задача Жеки

bookinisto commented on src/main/java/faang/school/postservice/service/like/LikeServiceImpl.java:
```diff
@@ -0,0 +1,151 @@
+        likeRepository.deleteByCommentIdAndUserId(commentId, userId);
+    }
+
+    private long getCurrentUserId() {
```
к Жеке

bookinisto commented on src/main/java/faang/school/postservice/service/like/LikeServiceImpl.java:
```diff
@@ -0,0 +1,151 @@
+        return userId;
+    }
+
+    private Post getPostOrThrow(long postId) {
```
к Жеке

bookinisto committed:
BJS2-84050
сделал :
- внес правки на коментарии

bookinisto committed:
Merge remote-tracking branch 'origin/murloc-master-stream11' into murloc-master-stream11

A

bookinisto committed:
Merge branch 'murloc-master-stream11' into murloc11/BJS2-84050-users-who-liked-it

bookinisto committed:
BJS2-84050
сделал :
- внес правки тестах : изменил пути url
- переименовал названия методов сделав короче

bookinisto committed:
BJS2-84050
сделал :
- внес правки тестах : изменил пути url
изменил "/likes/posts/{postId}" , postId
на "/likes/posts/" + postId

bookinisto committed:
BJS2-84050
сделал :
 внес правки тестах :
 - в LikeServiceImplTest исправил вызовы из ропозиториев поста и коментариев на getRequiredById
 - убрал создание объектов Optional в методах

RinatBeybutov commented on build.gradle.kts:
```diff
@@ -32,11 +32,6 @@ dependencies {
 
-    /**
-     * Kafka
-     */
```
почему удалил?

RinatBeybutov commented on src/main/java/faang/school/postservice/controller/comment/CommentController.java:
```diff
@@ -40,7 +40,7 @@
  */
 @RestController
-@RequestMapping("/posts/{postId}/comments")
+@RequestMapping("/comments")
```
а как будет postId передаваться?

RinatBeybutov commented on src/main/java/faang/school/postservice/controller/like/LikeController.java:
```diff
@@ -34,29 +41,39 @@
 public class LikeController {
 
-    private final LikeService service;
+    private final LikeService likeService;
```
какой смысл переименовывать было переменную?

RinatBeybutov commented on src/main/java/faang/school/postservice/mapper/CommentMapper.java:
```diff
@@ -50,7 +50,7 @@ public interface CommentMapper {
      * @return DTO представление комментария
      */
-    @Mapping(target = "likeCount", source = "likeCount")
+    @Mapping(target = "likeCount", expression = "java(entity.getLikes() != null ? (long) entity.getLikes().size() : 0L)")
```
зачем нужно было менять это обратно?

RinatBeybutov commented:
много удаленных файлов, с чем связано это удаление?

bookinisto committed:
Merge remote-tracking branch 'origin/murloc-master-stream11' into murloc-master-stream11

bookinisto committed:
пушу клон мастер для мерджа с рабочей веткой

bookinisto committed:
пушу клон мастер для мерджа с рабочей веткой

bookinisto committed:
Merge branch 'murloc11/BJS2-84050-users-who-liked-it' into murloc11/clonemaster11-2

bookinisto committed:
Merge pull request #2045 from CorporationX/murloc11/clonemaster11-2

Murloc11/clonemaster11 2

bookinisto committed:
BJS2-84050 user who liked it
сдела :  сделали с Евгением черипик мастер ветки на предидущий комит (откатили) , внеси изменения в мою рабочую ветку

RinatBeybutov commented on src/test/java/faang/school/postservice/controller/like/LikeControllerTest.java:
```diff
@@ -0,0 +1,93 @@
+    @MockBean
+    private UserContext userContext;
+
+    long userId = 1L;
```
private

RinatBeybutov replied:
private

RinatBeybutov commented on src/main/java/faang/school/postservice/service/like/LikeServiceImpl.java:
```diff
@@ -100,4 +109,35 @@ private long getCurrentUserIdAndValidate() {
+                });
+
+        return likes.stream()
+                .map(like -> userServiceClient.getUser(like.getUserId()))
```
у тебя получается здесь перебор в стриме и на каждый элемент рест запрос в юзер сервис, это может быть большая нагрузка на юзер сервис, которая его убьет.

А еще может быть ситуация, когда 99 запросов прошли, а 100й не пройдет и все рухнет. 

Здесь проще всего добавить эндпоинт на получение списка id пользователей в юзер сервисе.

RinatBeybutov commented on src/main/java/faang/school/postservice/service/like/LikeServiceImpl.java:
```diff
@@ -100,4 +109,35 @@ private long getCurrentUserIdAndValidate() {
+    public List<UserDto> getPostLikers(long postId) {
+        Post post = postRepository.getRequiredById(postId);
+
+        List<Like> likes = Optional.ofNullable(post.getLikes())
```
здесь может быть lazyInitException, нужно посмотреть повнимательнее на сущность. Посмотрю и к следующему ревью напишу, может и так можно оставить

bookinisto replied:
достаточно ли для решения будет добавить на метод getPostLikers @Transactional(readOnly = true) ??

RinatBeybutov commented on src/main/java/faang/school/postservice/service/like/LikeServiceImpl.java:
```diff
@@ -100,4 +109,35 @@ private long getCurrentUserIdAndValidate() {
+
+        List<Like> likes = Optional.ofNullable(post.getLikes())
+                .orElseGet(() -> {
+                    log.info("Post with id {} has no likes", postId);
```
давай здесь на уровень debug заменим, нечего такое в info писать, будет засорять лог

bookinisto committed:
BJS2-84050 user who liked it
сдела :
- в LikeControllerTest добавил к полям private
- сделал новый метод findAllByPostId в LikeRepository
- в LikeServiceImpl переписанны методы getPostLikers и getPostLikers
-в UserServiceClient добавлен метод getListUsers

bookinisto committed:
BJS2-84050 user who liked it
сдела :
мердж локальной ветки с мастер

bookinisto committed:
BJS2-84050 user who liked it
сделал :
мердж локальной ветки с мастер

bookinisto committed:
BJS2-84050 user who liked it
сделал :
разрешены конфликты в RedisConfig

RinatBeybutov commented on src/main/java/faang/school/postservice/client/UserServiceClient.java:
```diff
@@ -17,4 +19,7 @@ public interface UserServiceClient {
     List<UserDto> getUsersByIds(@RequestBody List<Long> ids);
+
+    @GetMapping("/users")
+    public List<UserDto> getListUsers(@RequestParam List<Long> userIds);
```
новый добавлять не нужно, используй тот, который в 21 строке

RinatBeybutov commented on src/main/java/faang/school/postservice/service/like/LikeServiceImpl.java:
```diff
@@ -100,4 +105,27 @@ private long getCurrentUserIdAndValidate() {
+    public List<UserDto> getPostLikers(long postId) {
+        List<Like> likes = likeRepository.findAllByPostId(postId);
+
+        List<Long> userIdList = likes.stream()
```
зашивать тип структуры данных в именование переменной - не самая хорошая идея. Лучше просто userIds

RinatBeybutov commented on src/main/java/faang/school/postservice/service/like/LikeServiceImpl.java:
```diff
@@ -100,4 +105,27 @@ private long getCurrentUserIdAndValidate() {
+                .toList();
+
+        return userServiceClient.getListUsers(userIdList);
+
```
лишний пробел

RinatBeybutov commented on src/main/java/faang/school/postservice/service/like/LikeServiceImpl.java:
```diff
@@ -100,4 +105,27 @@ private long getCurrentUserIdAndValidate() {
+
+    @Override
+    public List<UserDto> getCommentLikers(long commentId) {
+        List<Like> likes = likeRepository.findAllByPostId(commentId);
```
сомнительная идея искать среди комментариев по postId = commentId

RinatBeybutov commented on src/test/java/faang/school/postservice/service/like/LikeServiceImplTest.java:
```diff
@@ -0,0 +1,82 @@
+    @InjectMocks
+    private LikeServiceImpl likeService;
+
+    long postId = 1L;
```
1. Во-первых, забыл private
2. Во-вторых, давай вынесем тестовые данные в отдельный класс LikeServiceTestData

RinatBeybutov commented on src/test/java/faang/school/postservice/service/post/PostServiceImplTest.java:
```diff
@@ -18,6 +18,7 @@
 import org.mockito.InjectMocks;
 import org.mockito.Mock;
 import org.mockito.junit.jupiter.MockitoExtension;
+import org.springframework.beans.factory.annotation.Qualifier;
```
импорт не применяется

bookinisto committed:
BJS2-84050 user who liked it
сделал :
- убрал лишний метод из юзер-клинт
- переименовал некоторые поля
- переделал методы получения лайков относительно изменений в клиенте
- вынес из тестов поля со значениями в дата классы
- сделал правки в тестах относительно изменений в клиенте

bookinisto committed:
BJS2-84050 user who liked it
сделал :
- скоректировал использование полей дата классов для тестов в лайк контролле и лайк сервисе
